src/debug.c

```diff
@@ -2305,16 +2409,19 @@ void serverLogHexDump(int level, char *descr, void *value, size_t len) {
 /* =========================== Software Watchdog ============================ */
 #include <sys/time.h>
 
-void watchdogSignalHandler(int sig, siginfo_t *info, void *secret) {
+void sigalrmSignalHandler(int sig, siginfo_t *info, void *secret) {
 #ifdef HAVE_BACKTRACE
     ucontext_t *uc = (ucontext_t*) secret;
 #else
     (void)secret;
 #endif
-    UNUSED(info);
     UNUSED(sig);
 
-    serverLogFromHandler(LL_WARNING,"\n--- WATCHDOG TIMER EXPIRED ---");
+    /* SIGALRM can be sent explicitly to the process calling kill() to get the stacktraces,
+    or every watchdog_period interval. In the last case, si_pid is not set */
```

Comment:
did you get to this conclusion empirically? or by some docs?
i remember i made a similar assumption about SIGSEGV, and eventually concluded it was wrong and used `si_code`

Reply:
At first, I tried to use `si_code`. On macOS, both cases got the same code so I **tried** using `si_pid` and got the behavior described. 
https://man7.org/linux/man-pages/man2/sigaction.2.html
Here it is described in what cases `si_pid` is populated. It is not explicitly guaranteed that it is not populated if the signal is sent by `setitimer()`.
I'll add a test to ensure this is the behavior on all systems/if they decide to change something in the future.

Reply:
i remember that it seemed to work at first and then i started noticing in random places that it doesn't.
looked like it was just left uninitialized in some cases.
i don't think a test will help for "if they decide to change something in the future", our old versions will have bugs.
we can dig into the kernel code (at least in Linux), but i'd rather rely on docs and avoid that speculation.

did you say that si_code worked for you too?

Reply:
on macOS I got the same `si_code` for both sending `kill SIGALRM` and timer expiration. 

btw,:
https://linux.die.net/man/2/setitimer
> POSIX.1-2001, SVr4, 4.4BSD (this call first appeared in 4.2BSD). POSIX.1-2008 marks getitimer() and **setitimer()** obsolete, recommending the use of the POSIX timers API ([timer_gettime](https://linux.die.net/man/2/timer_gettime)(2), [timer_settime](https://linux.die.net/man/2/timer_settime)(2), etc.) instead.

The si_code generated using the recommended function ([timer_create](https://linux.die.net/man/2/timer_create)) is well-defined

> The si_code field of the siginfo_t structure will be set to **SI_TIMER**

Reply:
so you're saying that in current code that uses `settimer()` we can't rely on si_code (empirically, it doesn't work), and `si_pid` does seem to work.
but by shifting to POSIX.1-2001, we can resolve that and we can rely on si_code that's explicitly documented.
is that right?

@yossigo WDYT?

Reply:
correct

Reply:
@oranagra Seems like a good change - remove an obsolete call and get better/predictable results that are aligned with the documentation.

Reply:
i just wonder if POSIX.1-2001 is old enough to rely on, or if we already depend on it.
i don't want to introduce compatibility issues with exotic or old systems.

Reply:
@oranagra I think it's reasonable to rely on it.

Reply:
@yossigo @oranagra 
Update on this one:
It works fine on Linux but for some reason, and although it is supposed to conform to POSIX std, it is not supported on Mac (I tried it on my local machine)

see SO thread here
https://stackoverflow.com/questions/27473964/is-mac-os-x-posix-incompliant-timer-settime

I guess we'll stay with the original timer

Reply:
i'm ok with keeping the current code (since empirically, it works).
or we can (if the code looks clean enough) use timer_create on Linux, and fall back to setitimer on other platforms (or only on macos)

Reply:
ok, indeed it looks like it'll cause too much mess to support both.
if we did, i din't mind explicitly using linux for one, or explicitly excluding macos from the other.
but since the current one seems working, let's drop it for now.
we can always pick it up again in the future